### PR TITLE
Add WMTS multidimensional community module

### PR DESCRIPTION
**Pull request by nmco**
This pull request adds a new community module that implements WMTS multidimensional domain discovery extensions as proposed in this document http://demo.geo-solutions.it/share/wmts-multidim/wmts_multidim_geosolutions.html.

Mailing list discussion about this: http://osgeo-org.1560.x6.nabble.com/Adding-a-multidimensional-extension-to-WMTS-as-a-GeoServer-community-module-with-integrations-with-G-td5268826.html

**This pull request depends on the following GWC pull request**: https://github.com/GeoWebCache/geowebcache/pull/423

Some changes were made in WMS RasterTimeDimensionDefaultValueTest.java test class in order to allow the reuse of the time related multidimensional data.

### CODE REVIEW

File: src/community/wmts-multi-dimensional/pom.xml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.geoserver</groupId>
+        <artifactId>community</artifactId>
+        <version>2.10-SNAPSHOT</version>
+    </parent>
+    <groupId>org.geoserver.community</groupId>
+    <artifactId>wmts-multi-dimensional</artifactId>
```
Comment (aaime) [inline src/community/wmts-multi-dimensional/pom.xml]: All modules have names prefixed with gs-, this one should not be an exception. E.g., gs-wmts-multidim

Comment (nmco) [inline src/community/wmts-multi-dimensional/pom.xml]: good catch

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    /**
+     * Helper method that will extract the dimensions from a coverage type info.
+     */
+    private static List<Dimension> extractDimensions(WMS wms, LayerInfo layerInfo, CoverageInfo typeInfo) {
+        List<Dimension> dimensions = new ArrayList<>();
+        for (Map.Entry<String, Serializable> entry : typeInfo.getMetadata().entrySet()) {
+            String key = entry.getKey();
+            Serializable value = entry.getValue();
+            if (key.equals(ResourceInfo.TIME)) {
+                dimensions.add(new RasterTimeDimension(wms, layerInfo, Converters.convert(value, DimensionInfo.class)));
```
Comment (aaime) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java]: The code should check if the DimensionInfo is enabled. We can have layers that have had a dimension configured and no more enabled, but keep the settings to make sure we don't lose the edits.
In particular I have a layer that used to have a time dimension, that then got removed at the source level, and disabled in the UI, the current code tries to extract a domain that's not there causing a NPE:

````
05 set 12:36:34 ERROR [geowebcache.GeoWebCacheDispatcher] - null http://localhost:8080/geoserver/gwc/service/wmts
java.lang.NullPointerException
	at org.geoserver.catalog.util.ReaderDimensionsAccessor.getTimeDomain(ReaderDimensionsAccessor.java:152)
	at org.geoserver.gwc.wmts.dimensions.RasterTimeDimension.getDomainValues(RasterTimeDimension.java:37)
	at org.geoserver.gwc.wmts.dimensions.Dimension.getDomainValuesAsStrings(Dimension.java:180)
	at org.geoserver.gwc.wmts.MultiDimensionalExtension.encodeLayerDimension(MultiDimensionalExtension.java:226)
	at org.geoserver.gwc.wmts.MultiDimensionalExtension.encodeLayerDimensions(MultiDimensionalExtension.java:212)
	at org.geoserver.gwc.wmts.MultiDimensionalExtension.encodeLayer(MultiDimensionalExtension.java:137)
	at org.geowebcache.service.wmts.WMTSGetCapabilities.layer(WMTSGetCapabilities.java:453)
	at org.geowebcache.service.wmts.WMTSGetCapabilities.contents(WMTSGetCapabilities.java:395)
	at org.geowebcache.service.wmts.WMTSGetCapabilities.generateGetCapabilities(WMTSGetCapabilities.java:146)
	at org.geowebcache.service.wmts.WMTSGetCapabilities.writeResponse(WMTSGetCapabilities.java:90)
	at org.geowebcache.service.wmts.WMTSService.handleRequest(WMTSService.java:315)
	at org.geowebcache.service.wmts.WMTSService$$FastClassBySpringCGLIB$$e221a779.invoke(<generated>)
````

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/RasterElevationDimension.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import org.geotools.coverage.grid.io.GridCoverage2DReader;
+import org.opengis.filter.Filter;
+
+import java.io.IOException;
+import java.util.TreeSet;
+
+/**
+ * Represents an elevation dimension of a raster.
+ */
+public class RasterElevationDimension extends Dimension {
```
Comment (aaime) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/RasterElevationDimension.java]: This whole family of classes could use a toString()  implementation, it's harder to debug without

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java (hunk 2 above)
Comment (aaime) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java]: Btw, patch for this, I could not proceed with the review without fixing it:

````
diff --git a/src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java b/src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java
index 822fe86..a12c36c 100644
--- a/src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java
+++ b/src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java
@@ -39,11 +39,11 @@ public final class DimensionsUtils {
     private static List<Dimension> extractDimensions(WMS wms, LayerInfo layerInfo, FeatureTypeInfo typeInfo) {
         List<Dimension> dimensions = new ArrayList<>();
         DimensionInfo timeDimension = typeInfo.getMetadata().get(ResourceInfo.TIME, DimensionInfo.class);
-        if (timeDimension != null) {
+        if (timeDimension != null && timeDimension.isEnabled()) {
             dimensions.add(new VectorTimeDimension(wms, layerInfo, timeDimension));
         }
         DimensionInfo elevationDimension = typeInfo.getMetadata().get(ResourceInfo.ELEVATION, DimensionInfo.class);
-        if (elevationDimension != null) {
+        if (elevationDimension != null && elevationDimension.isEnabled()) {
             dimensions.add(new VectorElevationDimension(wms, layerInfo, elevationDimension));
         }
         return dimensions;
@@ -58,15 +58,21 @@ public final class DimensionsUtils {
             String key = entry.getKey();
             Serializable value = entry.getValue();
             if (key.equals(ResourceInfo.TIME)) {
-                dimensions.add(new RasterTimeDimension(wms, layerInfo, Converters.convert(value, DimensionInfo.class)));
+                final DimensionInfo dimension = Converters.convert(value, DimensionInfo.class);
+                if (dimension.isEnabled()) {
+                    dimensions.add(new RasterTimeDimension(wms, layerInfo, dimension));
+                }
             } else if (key.equals(ResourceInfo.ELEVATION)) {
-                dimensions.add(new RasterElevationDimension(wms, layerInfo, Converters.convert(value, DimensionInfo.class)));
+                final DimensionInfo dimension = Converters.convert(value, DimensionInfo.class);
+                if (dimension.isEnabled()) {
+                    dimensions.add(new RasterElevationDimension(wms, layerInfo, dimension));
+                }
             } else if (value instanceof DimensionInfo) {
-                DimensionInfo dimensionInfo = (DimensionInfo) value;
-                if (dimensionInfo.isEnabled()) {
+                final DimensionInfo dimension = Converters.convert(value, DimensionInfo.class);
+                if (dimension.isEnabled()) {
                     if (key.startsWith(ResourceInfo.CUSTOM_DIMENSION_PREFIX)) {
                         String dimensionName = key.substring(ResourceInfo.CUSTOM_DIMENSION_PREFIX.length());
-                        dimensions.add(new RasterCustomDimension(wms, layerInfo, dimensionName, dimensionInfo));
+                        dimensions.add(new RasterCustomDimension(wms, layerInfo, dimensionName, dimension));
                     }
                 }
             }
````

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/MultiDimensionalExtension.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+    private TileLayerDispatcher tileLayerDispatcher;
+
+    private final WMS wms;
+    private final Catalog catalog;
+
+    public MultiDimensionalExtension(WMS wms, Catalog catalog) {
+        this.wms = wms;
+        this.catalog = catalog;
+        tileLayerDispatcher = GeoServerExtensions.bean(TileLayerDispatcher.class);
```
Comment (aaime) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/MultiDimensionalExtension.java]: Confused, why is this not using constructor injection?

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java (hunk 2 above)
Comment (nmco) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java]: Correct ... for vector types I delegate the dimensions domain values retrieving on the WMS already existing methods (which correctly check if the dimension is enable or not) ... but for rasters I could not found methods for doing that in the WMS (in a cleaning way) and I forgot this use case ... thanks for the patch.

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/MultiDimensionalExtension.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        LayerInfo layerInfo = getLayerInfo(tileLayer, layerName);
+        // getting this layer dimensions along with its values
+        List<Dimension> dimensions = DimensionsUtils.extractDimensions(wms, layerInfo);
+        // let's see if we have a spatial limitation
+        ReferencedEnvelope boundingBox = (ReferencedEnvelope) conveyor.getParameter("bbox", false);
+        // add any domain provided restriction and set the bounding box
+        Filter filter = Filter.INCLUDE;
+        for (Dimension dimension : dimensions) {
+            Object restriction = conveyor.getParameter(dimension.getDimensionName(), false);
+            dimension.setBoundingBox(boundingBox);
```
Comment (aaime) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/MultiDimensionalExtension.java]: This way the DescribeDomains response contains the spatial domain only if it was requested. For vector feature and structured coverage grid readers it should be present regardless instead, and return only the bounds of the remaining features/granules (e.g., if I set the date to a range of few hours, the bbox for scattered satellite images could be pretty small). Anyways, I'd treat this one as a bug and fix it after the pull request has been merged.

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/MultiDimensionalExtension.java (hunk 4 above)
Comment (nmco) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/MultiDimensionalExtension.java]: No idea  this should indeed use constructor injection.

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/HistogramUtils.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    /**
+     * Helper method that creates buckets for a time domain based on the provided resolution. The returned tuple
+     * will contain the domain representation and the domain buckets.
+     */
+    private static Tuple<String, List<Range>> getTimeBuckets(TreeSet<?> domainValues, String resolution) {
+        Tuple<Date, Date> minMax = DimensionsUtils.getMinMax(domainValues, Date.class);
+        resolution = resolution != null ? resolution : TIME_DEFAULT_RESOLUTION;
+        Tuple<String, List<Date>> intervals = getDateIntervals(minMax, resolution);
+        int i = 0;
+        while (intervals.second.size() >= HISTOGRAM_MAX_THRESHOLD && i < MAX_ITERATIONS) {
```
Comment (aaime) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/HistogramUtils.java]: There is some logic error here, running a histogram request against time a domain looking like:

````
<Domains xmlns="http://demo.geo-solutions.it/share/wmts-multidim/wmts_multi_dimensional.xsd" xmlns:ows="http://www.opengis.net/ows/1.1">
<DimensionDomain>
<ows:Identifier>elevation</ows:Identifier>
<Domain>
0.0,200.0,400.0,600.0,800.0,1000.0,1200.0,1400.0,1600.0,1800.0,2000.0,3000.0,4000.0,5000.0,6000.0,7000.0,8000.0,9000.0
</Domain>
<Size>18</Size>
</DimensionDomain>
<DimensionDomain>
<ows:Identifier>time</ows:Identifier>
<Domain>
2016-02-23T03:00:00.000Z,2016-02-23T06:00:00.000Z,2016-02-23T09:00:00.000Z,2016-02-23T12:00:00.000Z,2016-02-23T15:00:00.000Z,2016-02-23T18:00:00.000Z,2016-02-23T21:00:00.000Z,2016-02-24T00:00:00.000Z,2016-02-24T03:00:00.000Z,2016-02-24T06:00:00.000Z,2016-02-24T09:00:00.000Z,2016-02-24T12:00:00.000Z,2016-02-24T15:00:00.000Z,2016-02-24T18:00:00.000Z,2016-02-24T21:00:00.000Z,2016-02-25T00:00:00.000Z,2016-02-25T03:00:00.000Z,2016-02-25T06:00:00.000Z,2016-02-25T09:00:00.000Z,2016-02-25T12:00:00.000Z,2016-02-25T15:00:00.000Z,2016-02-25T18:00:00.000Z,2016-02-25T21:00:00.000Z,2016-02-26T00:00:00.000Z,2016-02-26T03:00:00.000Z,2016-02-26T06:00:00.000Z,2016-02-26T09:00:00.000Z,2016-02-26T12:00:00.000Z,2016-02-26T15:00:00.000Z,2016-02-26T18:00:00.000Z,2016-02-26T21:00:00.000Z,2016-02-27T00:00:00.000Z,2016-02-27T03:00:00.000Z,2016-02-27T06:00:00.000Z,2016-02-27T09:00:00.000Z,2016-02-27T12:00:00.000Z,2016-02-27T15:00:00.000Z,2016-02-27T18:00:00.000Z,2016-02-27T21:00:00.000Z,2016-02-28T00:00:00.000Z,2016-02-28T03:00:00.000Z,2016-02-28T06:00:00.000Z,2016-02-28T09:00:00.000Z,2016-02-28T12:00:00.000Z,2016-02-28T15:00:00.000Z,2016-02-28T18:00:00.000Z,2016-02-28T21:00:00.000Z,2016-02-29T00:00:00.000Z
</Domain>
<Size>48</Size>
</DimensionDomain>
<DimensionDomain>
<ows:Identifier>REFERENCE_TIME</ows:Identifier>
<Domain>
2016-02-23T00:00:00.000Z,2016-02-24T00:00:00.000Z,2016-02-25T00:00:00.000Z,2016-02-26T00:00:00.000Z
</Domain>
<Size>4</Size>
</DimensionDomain>
</Domains>
````

Results in:

````
<Histogram xmlns="http://demo.geo-solutions.it/share/wmts-multidim/wmts_multi_dimensional.xsd" xmlns:ows="http://www.opengis.net/ows/1.1">
<ows:Identifier>time</ows:Identifier>
<Domain>
2016-02-23T03:00:00.000Z/2016-02-29T00:00:00.000Z/P1M
</Domain>
<Values>48</Values>
</Histogram>
````

I believe the call to getDateIntervals should have been something else, getDateIntervals always returns a "second" with just one element.

File: src/community/wmts-multi-dimensional/pom.xml (hunk 1 above)
Comment (nmco) [inline src/community/wmts-multi-dimensional/pom.xml]: fixed

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java (hunk 2 above)
Comment (nmco) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java]: fixed with proper tests cases and extended to vector types

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/RasterElevationDimension.java (hunk 3 above)
Comment (nmco) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/RasterElevationDimension.java]: fixed

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/MultiDimensionalExtension.java (hunk 4 above)
Comment (nmco) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/MultiDimensionalExtension.java]: fixed

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/HistogramUtils.java (hunk 6 above)
Comment (nmco) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/HistogramUtils.java]: default resolution set to 1 minute as discussed

**Comment by aaime** [comment]
Thank you, good to go. The space domain thing will be treated as a bug fix (it's a community module after all)

**Pull request merged**